### PR TITLE
feat(server): add closed socket detection to FGCI server

### PR DESCRIPTION
**Pull request by benoitdm-oslandia**
This feature improves feedback/cancel management during the map rendering to handle when the client closes the in/out socket. This can occur when the client (an http server as nginx) has a timeout shorter than the rendering time (fe. the rendering is very long due database issue). Without this feature, the http server may resent new requests to the qgis server which is busy and can not handle it.

For each new FCGI request, we start a new thread to monitor the incoming socket stream and detect its closing. If the socket closes before the rendering, we update a feedback object to cancel the rendering jobs. When the rendering is done, the thread is closed.

[X] should be backported to LTR

### CODE REVIEW

File: src/core/vector/qgsvectorlayerrenderer.cpp [hunk 1]
```diff
@@ -188,7 +188,7 @@ void QgsVectorLayerRenderer::setLayerRenderingTimeHint( int time )
 
 QgsFeedback *QgsVectorLayerRenderer::feedback() const
 {
-  return mFeedback.get();
+  return renderContext()->feedback();
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayerrenderer.cpp:191]: Can you explain why these changes are necessary? https://github.com/qgis/QGIS/blob/master/src/core/maprenderer/qgsmaprendererjob.cpp#L600 should take care of it already...

Comment (benoitdm-oslandia) [reply src/core/vector/qgsvectorlayerrenderer.cpp:191]: This is a remain of a test: we did not understand why the feedback was created in the `QgsVectorLayerRenderer` then set back to the job context instead of using the one of the job context. This looked like an inversion of responsibility. But later, we saw some renderer classes use an extended feedback class... 

I will revert this modifition

File: src/core/vector/qgsvectorlayerrenderer.cpp [hunk 2]
```diff
@@ -220,6 +221,11 @@ bool QgsVectorLayerRenderer::render()
   bool res = true;
   for ( const std::unique_ptr< QgsFeatureRenderer > &renderer : mRenderers )
   {
+    if ( ( mFeedback.get()->isCanceled() )
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayerrenderer.cpp:224]:
```suggestion
    if ( mFeedback->isCanceled() )
```

**Comment by pblottiere** [comment]
Hello @benoitdm-oslandia,

Thanks for your work :+1:. I'm going to take a look in the course of this week.

File: src/core/vector/qgsvectorlayerrenderer.cpp (hunk 2 above)
Comment (benoitdm-oslandia) [reply src/core/vector/qgsvectorlayerrenderer.cpp:224]: Fixed!

**Comment by benoitdm-oslandia** [comment]
I do not understand why the OGC tests are failing. I ran them on my laptop successfully following the same steps from the `.github/workflows/ogc.yml`. 

Any idea?

**Comment by benoitdm-oslandia** [comment]
I do not understand why the tests are failing! Can you provide me some enlightenment?

**Comment by pblottiere** [comment]
Hello @benoitdm-oslandia,

> I do not understand why the OGC tests are failing. I ran them on my laptop successfully

I pulled your branch and I confirm that OGC tests are green locally. For now I don't know what in your current implementation is misleading GitHub Actions... 

First I'm going to take some time to play with your PR.

File: src/server/qgsfcgiserverresponse.cpp [hunk 3]
```diff
@@ -23,15 +23,141 @@
 #include <fcgi_stdio.h>
 #include <QDebug>
 
+#include <unistd.h>
+#include "qgslogger.h"
+
+#ifdef Q_OS_LINUX
+#include <sys/types.h>
+#include <sys/socket.h>
```
Comment (pblottiere) [inline src/server/qgsfcgiserverresponse.cpp:31]: Would it be possible to use Qt sockets instead of Unix based sockets?

File: src/server/qgsfcgiserverresponse.cpp [hunk 4]
```diff
@@ -23,15 +23,141 @@
...
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#endif
+
+#ifdef Q_OS_LINUX
 //
-// QgsFcgiServerResponse
+// QgsFCGXStreamData copied from libfcgi FCGX_Stream_Data
 //
+typedef struct QgsFCGXStreamData
```
Comment (pblottiere) [inline src/server/qgsfcgiserverresponse.cpp:40]: It's probably a naive question but why is it necessary to redefine the structure in this case?

File: src/server/qgsfcgiserverresponse.h [hunk 5]
```diff
@@ -76,13 +103,21 @@ class SERVER_EXPORT QgsFcgiServerResponse: public QgsServerResponse
...
+
   private:
     QMap<QString, QString> mHeaders;
     QBuffer mBuffer;
     bool mFinished    = false;
     bool mHeadersSent = false;
     QgsServerRequest::Method mMethod;
     int mStatusCode = 0;
+
+    SocketMonitoringThread *mSocketMonitoringThread = nullptr;
```
Comment (pblottiere) [inline src/server/qgsfcgiserverresponse.h:119]: Can we use a unique_ptr?

File: src/server/qgsserverresponse.h [hunk 6]
```diff
@@ -195,6 +196,11 @@ class SERVER_EXPORT QgsServerResponse
      * Clear internal buffer
      */
     virtual void truncate() = 0;
+
+    /**
+     * Returns the socket feedback if any
+     */
```
Comment (pblottiere) [inline src/server/qgsserverresponse.h:202]: \since

File: src/server/services/wms/qgsmaprendererjobproxy.h [hunk 7]
```diff
@@ -53,8 +53,9 @@ namespace QgsWms
        * Sequential or parallel map rendering.
        * \param mapSettings Passed to MapRendererJob
        * \param image The resulting image
+       * \param feedback socket feedback
```
Comment (pblottiere) [inline src/server/services/wms/qgsmaprendererjobproxy.h:56]: since

**Comment by pblottiere** [comment]
@benoitdm-oslandia I have some questions but everything seems to work well locally (the rendering is canceled when the web server timeout is reached).

About the OGC tests, can you `cat` the QGIS Server logfile in the CI? It could give us some information.

File: src/server/qgsfcgiserverresponse.cpp (hunk 3 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.cpp:31]: The socket is provided by the fcgi lib and I do not know how to associate a Qt socket to a socket fd. By the way I need to set the SO_LINGER option on the socket and it is not available with QAbstractSocket::SocketOption

Comment (pblottiere) [reply src/server/qgsfcgiserverresponse.cpp:31]: I never used it but did you take a look to `QLocalSocket::setSocketDescriptor(...)`?

File: src/server/qgsfcgiserverresponse.h (hunk 5 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.h:119]: yes!

File: src/server/qgsfcgiserverresponse.cpp (hunk 4 above)
Comment (pblottiere) [reply src/server/qgsfcgiserverresponse.cpp:40]: Did you try something like `FCGX_Request *r = static_cast<FCGX_Request*>(FCGI_stdin->fcgx_stream->data);`?

**Comment by benoitdm-oslandia** [comment]
> About the OGC tests, can you cat the QGIS Server logfile in the CI? It could give us some information.

I have changed the `ogc.yml` and `Dockerfile` to be able to use the QGIS_DEBUG env var (`QGIS_DEBUG=5`). The only thing I saw is correct logs (many many logs) until a crash of the qgis server without any error log. It occured when or before calling this url `/qgisserver_wms130?ReQuEsT=GetCapabilities&VeRsIoN=100.0.0&SeRvIcE=WMS`. 

I will try to investigate a bit further by retrieving the core dumped and by using gdb.

Here is the log relevant part:

```log
qgis_server_deps | ------------------------------------------------
qgis_server_deps | "Accept" :  "text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2"
qgis_server_deps | "Host" :  "172.18.0.3"
qgis_server_deps | "User-Agent" :  "Java/1.8.0_171"
qgis_server_deps | src/core/qgsmessagelog.cpp:29 : (logMessage) [0ms] 2023-02-16T10:36:01 Server[0] Using configuration file path from environment: /data/teamengine_wms_130.qgs
qgis_server_deps | src/core/qgsmessagelog.cpp:29 : (logMessage) [0ms] 2023-02-16T10:36:01 Server[0] REQUEST:GetCapabilities
qgis_server_deps | src/core/qgsmessagelog.cpp:29 : (logMessage) [0ms] 2023-02-16T10:36:01 Server[0] SERVICE:WMS
qgis_server_deps | src/core/qgsmessagelog.cpp:29 : (logMessage) [0ms] 2023-02-16T10:36:01 Server[0] VERSION:1.3.0
qgis_server_deps | src/core/qgsmessagelog.cpp:29 : (logMessage) [0ms] 2023-02-16T10:36:01 Server[0] Using configuration file path from environment: /data/teamengine_wms_130.qgs
qgis_server_deps | src/core/project/qgsprojectversion.cpp:48 : (QgsProjectVersion) [0ms] Version is set to 1.3.0
qgis_server_deps | src/core/project/qgsprojectversion.cpp:48 : (QgsProjectVersion) [0ms] Version is set to 1.3.0
qgis_server_deps | src/core/qgsmessagelog.cpp:29 : (logMessage) [0ms] 2023-02-16T10:36:01 Server[1] Found WMS capabilities document in cache
qgis_server_deps | "Sent 1 blocks of 23762 bytes"
qgis_server_deps | src/core/qgsmessagelog.cpp:29 : (logMessage) [1ms] 2023-02-16T10:36:01 Server[0] Request finished in 2 ms
qgis_server_nginx | 172.18.0.4 - - [16/Feb/2023:10:36:01 +0000] "GET /qgisserver_wms130?REQUEST=GetCapabilities&VERSION=1.3.0&SERVICE=WMS HTTP/1.1" 200 23762 "-" "Java/1.8.0_171" "-"
qgis_server_nginx | 172.18.0.4 - - [16/Feb/2023:10:36:01 +0000] "GET /qgisserver_wms130?ReQuEsT=GetCapabilities&SeRvIcE=WMS HTTP/1.1" 200 23762 "-" "Java/1.8.0_171" "-"
qgis_server_nginx | 172.18.0.4 - - [16/Feb/2023:10:36:01 +0000] "GET /qgisserver_wms130?ReQuEsT=GetCapabilities&VeRsIoN=1.3.0&SeRvIcE=WMS HTTP/1.1" 200 23762 "-" "Java/1.8.0_171" "-"
Error: is_server_nginx | 2023/02/16 10:36:01 [error] 7#7: *1 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 172.18.0.4, server: _, request: "GET /qgisserver_wms130?ReQuEsT=GetCapabilities&VeRsIoN=100.0.0&SeRvIcE=WMS HTTP/1.1", upstream: "fastcgi://172.18.0.2:5555", host: "172.18.0.3"
qgis_server_nginx | 172.18.0.4 - - [16/Feb/2023:10:36:01 +0000] "GET /qgisserver_wms130?ReQuEsT=GetCapabilities&VeRsIoN=100.0.0&SeRvIcE=WMS HTTP/1.1" 502 174 "-" "Java/1.8.0_171" "-"
Error: is_server_nginx | 2023/02/16 10:36:01 [error] 7#7: *1 connect() failed (111: Connection refused) while connecting to upstream, client: 172.18.0.4, server: _, request: "GET /qgisserver_wms130?ReQuEsT=GetCapabilities&VeRsIoN=100.0.0&SeRvIcE=WMS HTTP/1.1", upstream: "fastcgi://172.18.0.2:5555", host: "172.18.0.3"
```

File: src/server/qgsfcgiserverresponse.cpp (hunk 3 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.cpp:31]: But I still need to set a specific option to the socket and I can not do that with the QLocalSocket. But I could set the option with the C lib and then use the QLocalSocket to send signal when the socket is disconnected. Is it what you have in mind?

File: src/server/qgsfcgiserverresponse.cpp (hunk 4 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.cpp:40]: no because `data` is a QgsFCGXStreamData (originally found [here](https://github.com/FastCGI-Archives/fcgi2/blob/874b148dfeadf8ebadff89cf228e63f836e509a3/libfcgi/fcgiapp.c#L1333)) disguised in a `void *`

File: src/server/qgsserverresponse.h (hunk 6 above)
Comment (benoitdm-oslandia) [reply src/server/qgsserverresponse.h:202]: Done!

File: src/server/services/wms/qgsmaprendererjobproxy.h (hunk 7 above)
Comment (benoitdm-oslandia) [reply src/server/services/wms/qgsmaprendererjobproxy.h:56]: Done!

**Comment by benoitdm-oslandia** [comment]
> About the OGC tests, can you cat the QGIS Server logfile in the CI? It could give us some information.

I have modified the build and `ogc.yml` file to activate the DEBUG in the cmake and ran it with `QGIS_DEBUG=5` but the error message I got is not very useful (see https://github.com/benoitdm-oslandia/QGIS/actions/runs/4222576096/jobs/7331242146#step:9:1055):

```log
qgis_server_deps | src/core/project/qgsprojectversion.cpp:48 : (QgsProjectVersion) [0ms] Version is set to 1.3.0
qgis_server_deps | src/core/qgsmessagelog.cpp:29 : (logMessage) [0ms] 2023-02-20T10:55:10 Server[1] Set WMS capabilities document in cache
qgis_server_deps | "Sent 1 blocks of 23762 bytes"
qgis_server_deps | src/core/qgsmessagelog.cpp:29 : (logMessage) [6ms] 2023-02-20T10:55:10 Server[0] Request finished in 162 ms
qgis_server_deps | QThread: Destroyed while thread is still running
```

I have retrieve the core dump and read it with `gdb` without any clues (see: https://github.com/benoitdm-oslandia/QGIS/actions/runs/4222576096/jobs/7331242146#step:9:333:

```log
[New LWP 1]
[New LWP 9]
[New LWP 10]
[New LWP 11]
[New LWP 12]
Core was generated by `/usr/src/qgis/build/output/bin/qgis_mapserv.fcgi'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f0a91269898 in ?? ()
[Current thread is 1 (LWP 1)]
(gdb) #0  0x00007f0a91269898 in ?? ()
#1  0x0000000000000020 in ?? ()
#2  0x0000000000000000 in ?? ()
(gdb) quit
Attaching to qgis_server_nginx, qgis_server_deps
```

**Comment by benoitdm-oslandia** [comment]
I  found the pbm in the `qgsfcgiserverresponse` destructor!

**Comment by pblottiere** [comment]
Hello @benoitdm-oslandia,

Sorry for not responding, I was in vacation last week. I'm going to take a look in the coming days, especially on https://github.com/qgis/QGIS/pull/51744#pullrequestreview-1301366358 and https://github.com/qgis/QGIS/pull/51744#pullrequestreview-1301367738.

**Comment by benoitdm-oslandia** [comment]
@pblottiere any news?

**Comment by rldhont** [comment]
@benoitdm-oslandia can you add cancelation into WFS request ?

Would it be possible to cancel from Python and not only fcgi ?

**Comment by benoitdm-oslandia** [comment]
> @benoitdm-oslandia can you add cancelation into WFS request ?

hmmm may be ...  WMS can be slow on rendering and thus need to be interruptable, in what case do you have slow responses with WFS that need to be interrupted?

> Would it be possible to cancel from Python and not only fcgi ?

I think (but I am not sure) this is already possible from python.

**Comment by benoitdm-oslandia** [comment]
Hi @pblottiere! Can you have a look at this PR please?

**Comment by pblottiere** [comment]
Hello @benoitdm-oslandia,

> Can you have a look at this PR please?

Sorry for the delay, I'm taking a look.

File: src/server/qgsfcgiserverresponse.cpp [hunk 8]
```diff
@@ -23,15 +23,148 @@
...
+  : mIsResponseFinished( isResponseFinished )
+  , mFeedback( feedback )
+  , mIpcFd( -1 )
+{
+  setObjectName( "FCGI socket monitor" );
+  Q_ASSERT( mIsResponseFinished );
+  Q_ASSERT( mFeedback );
+
+#ifdef Q_OS_LINUX
+  if ( FCGI_stdout && FCGI_stdout->fcgx_stream && FCGI_stdout->fcgx_stream->data )
```
Comment (pblottiere) [inline src/server/qgsfcgiserverresponse.cpp:71]: I think that we should check for the `FCGI_stdin` stream in this case (not `FCGI_stdout`).

**Comment by pblottiere** [comment]
@benoitdm-oslandia I took some time for testing your PR and it works well :+1:. Such a feature could be very useful.

However, as you may have already noticed, I'm a little hesitant with this PR, mainly because of portability and stability issues. Basically, my concerns are:

- it's not cross platform due to the `setsockopt` syscall
- it's clearly stated in FCGI documentation that `ipcFd` should not be used   (internal API only)

Moreover, I noticed that the `FCGI_stdin` socket is configured with `SO_LINGER`. Such an option results in sending a RST packet to QGIS Server when the Web server timeout is reached. With the current implementation, I think that it's not useful because the `recv` call is made on the `FCGI_stdin` side, so the RST packet sent to QGIS Server is not really detected/used.

So I just tried to deactivate the `setsockopt` call and judging from what I've seen, the `recv` function is still returning 0 when the timeout is reached.

@benoitdm-oslandia please let me know if I missed something.

@elpaso @dmarteau can you give me your opinion on this PR too please?

File: src/server/qgsfcgiserverresponse.cpp (hunk 8 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.cpp:71]: The fact is the qgis server will need to write the response to the FCGI_stdout. This is why I monitor this stream

**Comment by benoitdm-oslandia** [comment]
> it's clearly stated in FCGI documentation that ipcFd should not be used (internal API only)

Yes, but I need this fd to monitor the socket. I could ask them to add some getter function to retrieve the data or provide another mechanism but it will take a while to have all OS update their fcgi lib. Also the `ipcFd` has been defined 24 years ago, I do not think it will ever change :)

> it's not cross platform due to the setsockopt syscall

Yes, I agree and I am annoyed to be unable to find another cross-plateform solution: Qt does not support the option and there is no existing class/function to do the job. But may be I missed it! 

> So I just tried to deactivate the setsockopt call and judging from what I've seen, the recv function is still returning 0 when the timeout is reached.

hum... for what I tried, it did not work without it! I will check it again. It may be linked to the OS or the fcgi spawn.

**Comment by benoitdm-oslandia** [comment]
@pblottiere can you have a look please?

**Comment by pblottiere** [comment]
Hello @benoitdm-oslandia,

Sorry for the delay, I'll take a look friday.

And did you notice that OGC tests are failing?

**Comment by benoitdm-oslandia** [comment]
Yes they still falling in the GitHub CI but they are running fine on my PC.

**Comment by pblottiere** [comment]
Hello @benoitdm-oslandia,

> I have finally found the solution with QLocalSocket: it allows me to reduce the part of code dedicated to linux to a smaller part and remove some hacks!

Nice! I took a look and I'm much happier, especially now that the `SO_LINGER` option is removed :sweat_smile:.

However I pulled your branch and I have some strange behavior (segfault...). I'm going to investigate (it may be due to my last upgrade...).

**Comment by pblottiere** [comment]
Hello @benoitdm-oslandia,

> However I pulled your branch and I have some strange behavior (segfault...). I'm going to investigate (it may be due to my last upgrade...).

I performed some more tests on your branch and I noticed two issues.

First, exceptions are not forwarded to client anymore (sockets are failing?). For example with your branch:

```` console
$ curl http://localhost/qgisserver
<!DOCTYPE html>
<html>
<head>
<title>Error</title>
<style>
html { color-scheme: light dark; }
body { width: 35em; margin: 0 auto;
font-family: Tahoma, Verdana, Arial, sans-serif; }
</style>
</head>
<body>
<h1>An error occurred.</h1>
<p>Sorry, the page you are looking for is currently unavailable.<br/>
Please try again later.</p>
<p>If you are the system administrator of this resource then you should check
the error log for details.</p>
<p><em>Faithfully yours, nginx.</em></p>
</body>
</html>
````

While on master, we have a valid exception:

```` console
$ curl http://localhost/qgisserver
<?xml version="1.0" encoding="UTF-8"?>
<ServiceExceptionReport xmlns="http://www.opengis.net/ogc" version="1.3.0">
 <ServiceException code="Service configuration error">Service unknown or unsupported. Current supported services (case-sensitive): WMS WFS WCS WMTS SampleService, or use a WFS3 (OGC API Features) endpoint</ServiceException>
</ServiceExceptionReport>
````

BTW, I noticed the `QObject::connect: Cannot queue arguments of type 'QAbstractSocket::SocketError'` message in QGIS Server logs, so it's probably relevant.


Morever, I noticed a segmentation fault when:

1. two `GetMap` requests are simultaneously sent to QGIS Server whith a  rendering time exceeding the HTTP Server timeout
2. the first `GetMap` request is processed normally 
4. then a segmentation fault is raised before proceeding the second request  with the next error: `QSocketNotifier: Socket notifiers cannot be enabled or disabled from another thread`

(the previous error is easily reproducible with `fastcgi_read_timeout 3;` in NGINX configuration and a simple 10 seconds sleep in `QgsRenderer::getMap()` method)

**Comment by benoitdm-oslandia** [comment]
Hi @pblottiere, I just retried with my previous version (the more low level version, not the Qt one) and it worked fine!

It looked for some workarounds for the Qt version bug (`QObject::connect: Cannot queue arguments of type 'QAbstractSocket::SocketError'`), but none worked. Do you have any idea?

**Comment by benoitdm-oslandia** [comment]
I had `qRegisterMetaType<QAbstractSocket::SocketError>("QAbstractSocket::SocketError");` at the beginning of my `SocketMonitoringThread::run()`. I have no crash but I cannot retrieve the correct error message. It is like the `QLocalSocket` is taking the ownership of the output socket :disappointed:

**Comment by benoitdm-oslandia** [comment]
@pblottiere I hope this will be the good one!

**Comment by benoitdm-oslandia** [comment]
@pblottiere could review it again please?

**Comment by dmarteau** [comment]
Hi,

This feature is undoubtedly useful but why not exposing the  `virtual QgsFeedback *feedback() const`  method in the python api ?

Imho, the idea to propagate a cancellation method from the connection state  is a general enough concept so that it may be used by any other QgsServerResponse implementor using other channel than  FCGI.

**Comment by benoitdm-oslandia** [comment]
> Hi,
> 
> This feature is undoubtedly useful but why not exposing the `virtual QgsFeedback *feedback() const` method in the python api ?
> 
> Imho, the idea to propagate a cancellation method from the connection state is a general enough concept so that it may be used by any other QgsServerResponse implementor using other channel than FCGI.

Done!

**Comment by benoitdm-oslandia** [comment]
@pblottiere @rldhont can you review this PR please?

**Comment by rldhont** [comment]
@benoitdm-oslandia I am not able to review the fcgi part. I approve the rest of the code.

File: src/server/qgsfcgiserverresponse.cpp [hunk 9]
```diff
@@ -23,15 +23,138 @@
 #include <fcgi_stdio.h>
 #include <QDebug>
 
+#include <unistd.h>
+#include "qgslogger.h"
+
+#ifdef Q_OS_LINUX
```
Comment (lbartoletti) [inline src/server/qgsfcgiserverresponse.cpp:29]: Q_OS_UNIX for all unix-like including bsd? ping @landryb

File: src/server/services/wms/qgsmaprendererjobproxy.h [hunk 10]
```diff
@@ -53,8 +53,10 @@ namespace QgsWms
...
+       * \since 3.30.0
```
Comment (lbartoletti) [inline src/server/services/wms/qgsmaprendererjobproxy.h:57]: You can bump it :)

File: src/server/services/wms/qgswmsrendercontext.h [hunk 11]
```diff
@@ -270,6 +270,18 @@ namespace QgsWms
        */
       bool isExternalLayer( const QString &name ) const;
 
+      /**
+       * Sets the response feedback.
+       * \since QGIS 3.30
```
Comment (lbartoletti) [inline src/server/services/wms/qgswmsrendercontext.h:275]: same

File: src/server/services/wms/qgswmsrendercontext.h [hunk 12]
```diff
@@ -306,6 +318,8 @@ namespace QgsWms
 
       // list for external layers
       QList<QgsMapLayer *> mExternalLayers;
+
+      QgsFeedback *mSocketFeedback = nullptr;
```
Comment (lbartoletti) [inline src/server/services/wms/qgswmsrendercontext.h:322]: Is a smart pointers possible here?

File: src/server/qgsfcgiserverresponse.cpp (hunk 9 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.cpp:29]: Should I put a `Q_OS_LINUX && Q_OS_UNIX`?

File: src/server/services/wms/qgsmaprendererjobproxy.h (hunk 10 above)
Comment (benoitdm-oslandia) [reply src/server/services/wms/qgsmaprendererjobproxy.h:57]: done!

File: src/server/services/wms/qgswmsrendercontext.h (hunk 11 above)
Comment (benoitdm-oslandia) [reply src/server/services/wms/qgswmsrendercontext.h:275]: done!

File: src/server/qgsfcgiserverresponse.cpp (hunk 9 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.cpp]: I made a mix

**Comment by benoitdm-oslandia** [comment]
@lbartoletti all tests are green, can we say we merge this PR by the end of week if there is no other comment?

File: src/server/qgsfcgiserverresponse.h [hunk 13]
```diff
@@ -26,6 +26,38 @@
...
 #include <QBuffer>
+#include <QThread>
+#include <QLocalSocket>
+
+
+/**
+ * \ingroup server
+ * \class SocketMonitoringThread
+ * \brief Thread used to monitor the fcgi socket
+ * \since QGIS 3.30
```
Comment (Gustry) [inline src/server/qgsfcgiserverresponse.h:37]: Version ?

File: src/server/qgsfcgiserverresponse.h [hunk 14]
```diff
@@ -76,13 +109,22 @@ class SERVER_EXPORT QgsFcgiServerResponse: public QgsServerResponse
      */
     void setDefaultHeaders();
 
+    /**
+     * Returns socket feedback if any
+     * \since QGIS 3.30
```
Comment (Gustry) [inline src/server/qgsfcgiserverresponse.h:114]: same ?

File: src/server/qgsfcgiserverresponse.h (hunk 13 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.h]: fixed!

**Comment by benoitdm-oslandia** [comment]
thanks @Gustry!

**Comment by pblottiere** [comment]
Hello @benoitdm-oslandia,

I pulled your branch but unfortunately I still have really weird behaviors. One-in-two request returns an error:

``` console
<!DOCTYPE html>
<html>
<head>
<title>Error</title>
<style>
html { color-scheme: light dark; }
body { width: 35em; margin: 0 auto;
font-family: Tahoma, Verdana, Arial, sans-serif; }
</style>
</head>
<body>
<h1>An error occurred.</h1>
<p>Sorry, the page you are looking for is currently unavailable.<br/>
Please try again later.</p>
<p>If you are the system administrator of this resource then you should check
the error log for details.</p>
<p><em>Faithfully yours, nginx.</em></p>
</body>
</html>
```

And I have the next messages in QGIS Server logs:

``` console
QSocketNotifier: Invalid socket 16 and type 'Read', disabling...
QObject::connect: Cannot queue arguments of type 'QAbstractSocket::SocketState'
(Make sure 'QAbstractSocket::SocketState' is registered using qRegisterMetaType().)
```

> all tests are green, can we say we merge this PR by the end of week if there is no other comment?

This MR is not ready yet. And I can't stress enough that this MR can be very harmful for QGIS Server in case of a buggy implementation, so we need to be cautious.

**Comment by benoitdm-oslandia** [comment]
@pblottiere I have been able to validate it without the LINGER flag and without QLocalSocket. 

Can you have a look please?
